main/models/space_membership.py

```diff
@@ -0,0 +1,290 @@
+            logger.exception("Error in send_approval_request - failed to send email",
+                             extra={'SpaceMembership': self})
+
+    # email space to notify of decision
+    def send_application_decision(self):
+        htmly = get_template('join_space/space_decision_email.html')
+
+        d = {
+            'email': self.applied_by.email,
+            'first_name': self.applied_by.first_name,
+            'last_name': self.applied_by.last_name,
+            'space_name': self.space.name,
+            'fee': self.fee,
+            'status': self.status
+        }
+
+        subject = "Hackspace Foundation Membership Application"
+        from_email = getattr(settings, "BOARD_EMAIL", None)
+        to = self.applied_by.email
+        cc = self.space.email
+        message = htmly.render(d)
+        try:
+            msg = EmailMessage(subject, message, to=[to], cc=[cc], from_email=from_email)
+            msg.content_subtype = 'html'
+            msg.send()
+
+            return True
+        except Exception:
+            logger.exception("Error in send_application_decision - unable to send email",
+                             extra={'membership application': self})
+            return False
+
+    # approve the membership application (and create initial payment)
+    def approve(self):
+        # check space has not already been approved/rejected (e.g. by someone else!)
+        if self.status != 'Pending':
+            return False
+
+        # update membership status
+        self.status = 'Approved'
+        self.save()
+
+        self.send_application_decision()
+
+        self.request_payment()
+
+        return True
+
+    # reject the membership application (and cancel mandate)
+    def reject(self):
+        # check space has not already been approved/rejected (e.g. by someone else!)
+        if self.status != 'Pending':
+            return False
+
+        # update membership status
+        self.status = 'Rejected'
+        self.save()
+
+        self.send_application_decision()
+
+        if self.has_active_mandate():
+            return self.mandate().cancel()
+
+        return True
+
+    # request new payment for this membership (e.g. start of a new year)
+    def request_payment(self):
+        if self.has_active_mandate():
+            return self.mandate().create_payment(self.fee)
+
+    def handle_payment_received(self, payment):
+        if payment.payout_date is not None:
+            # update started_at when first payment received
+            self.started_at = payment.payout_date
+
+            # update expired_at when new payment received
+            self.expired_at = payment.payout_date + timedelta(days=365)
+
+            self.save()
+        else:
+            logger.error("handle_payment_received - payout_date is null")
+
+        # TODO: send notification email of payment received and membership active
+
+    def handle_mandate_updated(self, mandate):
+        # TODO: something useful
```

Comment:
Presuming this is only for data like chanced bank details, cancelled mandates and won't effect normal operation not to process this data?

Reply:
Stole these from supporter_membership which is also still a TODO